Build failure: pcalc

### Nixpkgs version

- Unstable (26.05)

### Steps to reproduce

`nix build .#pcalc` on `master`

### Can Hydra reproduce this build failure?

Yes, Hydra can reproduce this build failure.

### Link to Hydra build job

https://hydra.nixos.org/build/317885403

### Relevant log output

```console
hocdecl.h:45:13: note: 'echo' declared here
   45 | extern  int echo(const char *str);
      |             ^~~~
symbol.c:99:22: error: initialization of 'int (*)(void)' from incompatible pointer type 'int (*)(const char *)' [-Wincompatible-pointer-types]
   99 |         {"_ECHO",    echo_nl},
      |                      ^~~~~~~
symbol.c:99:22: note: (near initialization for 'ibuiltins[7].ifunc')
hocdecl.h:46:13: note: 'echo_nl' declared here
   46 | extern  int echo_nl(const char *str);
      |             ^~~~~~~
make: *** [<builtin>: symbol.o] Error 1
make: *** Waiting for unfinished jobs....
```

### Additional context

_No response_

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.12.60, NixOS, 26.05 (Yarara), 26.05.20251205.f61125a`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.31.2`
 - nixpkgs: `/nix/store/cd6v80pj9h7r2prxx9a71y92zq88p1xx-source`

### Notify maintainers

@ftrvxmtrx

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+label%3A%220.kind%3A+build+failure%22).
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/pc/pcalc/package.nix
+++ b/pkgs/by-name/pc/pcalc/package.nix
@@ -17,6 +17,11 @@ stdenv.mkDerivation {
     sha256 = "sha256-m4xdsEJGKxLgp/d5ipxQ+cKG3z7rlvpPL6hELnDu6Hk=";
   };
 
+  # Since C23, coercing functions with different parameter lists to a function pointer with no
+  # parameter specified triggers a hard error: `symbol.c:92:22: error: initialization of
+  # 'int (*)(void)' from incompatible pointer type 'int (*)(double)' [-Wincompatible-pointer-types]`
+  env.NIX_CFLAGS_COMPILE = "-std=gnu17";
+
   makeFlags = [ "DESTDIR= BINDIR=$(out)/bin" ];
   nativeBuildInputs = [
     bison

